src/Neo.VM/ReferenceCounter.cs

```diff
@@ -146,10 +148,17 @@ internal void RemoveReference(StackItem item, CompoundType parent)
 
         internal void RemoveStackReference(StackItem item)
         {
-            references_count--;
+            DecreaseCounter();
             if (!NeedTrack(item)) return;
             if (--item.StackReferences == 0)
                 zero_referred.Add(item);
         }
+
+        private void DecreaseCounter(int count = 1)
+        {
+            references_count -= count;
+            if (references_count < 0) throw new InvalidOperationException("Reference counter is negative.");
```

Comment:
But how is it ever possible? Have you faced with the issue when refcount is negative?

Reply:
Yes, its @dusmart who found the issue. With Array object deepcopied and removed from the stack, all of its references are removed as well, but when we add that array back to the stack, its internal subitems will not be added to the reference counter, but when we remove that item from the stack again, the counter will consider the array subitems.

BArray = AArray.DeepCopy

Remove BArray from stack // this will remove BArray and its subitem from the counter. 

Aadd BArray back to stack // This will only add BArray itself back to the counter

Remove BArray from stack // This will remove BArray and its subitems from the counter, and create minus counter.

Reply:
AArray.DeepCopy doesn't increase the counter?

Reply:
> AArray.DeepCopy doesn't increase the counter?

It should increase the counter:
https://github.com/neo-project/neo/blob/6e8c7308267b046fd24fa232f5d0ea35f6a07b5d/src/Neo.VM/Types/Array.cs#L75

Reply:
>  if (references_count < 0) throw new InvalidOperationException("Reference counter is negative.");

I think we should not change this code. It's a bug in the handlers implementation, not in the reference counter logic. So we should fix the source problem, not its consequence.

Reply:
the deepcopy happens off the stack, the result is not in the stack, thus the counter will be increased during the deepcopy, but decreased right after its assigned (or after the first time of checking the counter).

Reply:
> 2. reference counter is supposed to be not negative, it should be the counter's property, should not rely on the callers.

@Jim8y let's discuss it here, we have a designated thread for it. If we add this change, then it may affect some old transactions. And I'd like to see @roman-khimov's opinion on this topic.

Reply:
Why? isn't you saying its supposed to be positive? then what problem can this change cause?

Reply:
i suggest to move the reference counter into another pr and keep the fix simple